client/src/components/EmojiPicker.js

```diff
@@ -0,0 +1,169 @@
+    .map((e, index) => <li key={index}
+                           onMouseEnter={ e => { e.preventDefault(); onMouseEnter(index); }}
+                           onClick={ e => { e.preventDefault(); onClick(index); }}
+                           className={index === highlightedIndex ? 'selected' : ''}>{e}</li>);
+};
+
+const EmojiList = ({
+    emojis,
+    highlightedIndex,
+    onMouseEnter,
+    onClick
+}) => <ul> { emojify(emojis, highlightedIndex, onMouseEnter, onClick) } </ul>
+
+class EmojiPicker extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            emojis: [],
+            highlightedIndex: 0,
+            keysDown: {}
+        }
+    }
+
+    selectEmoji(index) {
+        const emojiShortName = this.state.emojis[index] ? this.state.emojis[index].shortname : this.props.filterText;
+        this.props.onSelectEmoji(emojiShortName);
+    }
+
+    onKeyUp(event) {
+        if (event.which === 16) {
+            // Shift
+            const keysDown = { 'shift' : false };
+            this.setState({ keysDown: keysDown });
+        }
+    }
+    onKeyDown(event) {
+        if (event.which === 8) {
+            // Backspace
+            this.selectEmoji(-1);
+            if(this.state.highlightedIndex !== 0) {
+                event.preventDefault();
+                this.setState({ highlightedIndex: 0});
+            }
+        } else if (event.which === 16) {
+            // Shift
+            const keysDown = { 'shift' : true };
+            this.setState({ keysDown: keysDown });
+        } else if(event.which === 9) {
+            // Tab
+            event.preventDefault();
+            if (this.state.keysDown['shift']) {
+                // Shift + Tab
+                this.cycle(-1);
+            } else {
+                this.cycle(1);
+            }
+        } else if (event.which === 39) {
+            // Right arrow
+            event.preventDefault();
+            this.cycle(1);
+        } else if (event.which === 37) {
+            // Left arrow
+            event.preventDefault();
+            this.cycle(-1);
+        } else if (event.which === 38) {
+            // Up arrow
+            event.preventDefault();
+            const step = -this.props.elemsPerRow;
+            this.cycleRow(step);
+        } else if (event.which === 40) {
+            // Down arrow
+            event.preventDefault();
+            const step = this.props.elemsPerRow;
+            this.cycleRow(step);
+        } else if (event.which === 13 || event.which === 32) {
+            // Return or Space
+            event.preventDefault();
+            this.selectEmoji(this.state.highlightedIndex);
+            this.props.onClose();
+        } else if (event.which === 27) {
+            //ESC
+            this.selectEmoji(-1);
+            this.props.onClose();
+          }
+      }
+
+      onClick(index) {
+          this.setState({ highlightedIndex: index});
+          this.selectEmoji(index);
```

Comment:
I think you can remove the two lines here. They get called at `onMouseEnter`. If you do that, you need to add `      this.refs.message.focus();` to `onCloseEmojiPicker()` in SendMessage.js.